src/mongo/client/insert_write_operation.cpp

```diff
@@ -66,8 +67,11 @@ namespace mongo {
     }
 
     BSONObj InsertWriteOperation::_ensureId(const BSONObj& doc) {
-        if (doc.hasField("_id"))
+        BSONElement id = doc.getField("_id");
+        if (!id.eoo()) {
+            uassert(0, "value of _id element cannot contain any fields starting with $", !id.isABSONObj() || id.Obj().okForStorage());
```

Comment:
Does isABSONObj also handle the case where _id is an array (I don't know that is possible, btw).